### PR TITLE
Support reading podman logs

### PR DESCRIPTION
**Pull request by djmitche**
This is controlled by a config flag as a temporary measure until we have a more refined approach in autodiscovery.

* Look in a podman-specific directory
* Parse logs in the format podman writes (kubernetes)

### Motivation

Get basic support for podman logs in the 7.33.0 release.

### Additional Notes

This PR is currently based on #9588 and thus shouldn't be reviewed until that lands, at which point I will rebase this PR and it will contain only a single commit.

### Describe how to test your changes

Run the containerized agent in podman to enable AD to identify podman.  Note that podman can be installed on a mac (brew install podman) or in a VM.  But the latest Ubuntu versions of podman [are broken](https://github.com/kubic-project/issues/issues/15).

```
[core@localhost ~]$ API_KEY=...
[core@localhost ~]$ HOST=...
[core@localhost ~]$ IMAGE=...
[core@localhost ~]$ sudo podman run -d \
     --name dd-agent \
     -v /run/podman/podman.sock:/run/podman/podman.sock:ro \
     -v /proc/:/host/proc/:ro \
     -v /sys/fs/cgroup/:/host/sys/fs/cgroup:ro \
     -e DD_API_KEY=$API_KEY \
     -e DOCKER_HOST=unix:///run/podman/podman.sock \
     --privileged \
     -e DD_LOGS_ENABLED=true \
     -e DD_LOGS_CONFIG_CONTAINER_COLLECT_ALL=true \
     -e DD_HOSTNAME=$HOST \
     -e DD_LOGS_CONFIG_USE_PODMAN_LOGS=true \
     $IMAGE
...
[core@localhost ~]$ sudo podman run -d bash -c 'while true; do date; sleep 1; done'
..
[core@localhost ~]$ sudo podman ps
CONTAINER ID  IMAGE                                                    COMMAND               CREATED         STATUS                       PORTS       NAMES
a7738ce94558  docker.io/datadog/agent-dev:dustin-mitchell-ac-1054-py3  /bin/entrypoint.s...  39 seconds ago  Up 38 seconds ago (healthy)              dd-agent
dd7ad06a44e6  docker.io/library/bash:latest                            -c while true; do...  4 seconds ago   Up 4 seconds ago                         inspiring_shaw
[core@localhost ~]$ sudo podman exec -ti dd-agent agent stream-logs
...
```

### CODE REVIEW

File: pkg/logs/input/file/tailer.go [hunk 1]
```diff
@@ -76,7 +76,13 @@ func NewDecoderFromSourceWithPattern(source *config.LogSource, multiLinePattern
...
 		matcher = &decoder.NewLineMatcher{}
 	case config.DockerSourceType:
-		lineParser = parser.DockerFileFormat
+		switch coreConfig.Datadog.GetBool("logs_config.use_podman_logs") {
+		case false:
+			lineParser = parser.DockerFileFormat
+		case true:
+			// podman's on-disk logs are in kubernetes format
+			lineParser = parser.KubernetesFormat
+		}
```
Comment (sgnn7) [inline pkg/logs/input/file/tailer.go:85]: nit: I'm a bit curious about use of a `switch` here for a boolean value. Any reason why we wouldn't use an `if`/`else` or just an `if`?

```go
lineParser = parser.DockerFileFormat
if coreConfig.Datadog.GetBool("logs_config.use_podman_logs") {
    lineParser = parser.KubernetesFormat
}
```

File: pkg/logs/input/docker/launcher.go [hunk 2]
```diff
@@ -298,10 +299,15 @@ func (l *Launcher) getFileSource(container *Container, source *config.LogSource)
...
-	return filepath.Join(basePath, id, filename)
+func getPath(id string) string {
+	var tpl string
+	switch coreConfig.Datadog.GetBool("logs_config.use_podman_logs") {
+	case false:
+		tpl = filepath.Join(basePath, "$ID/$ID-json.log")
+	case true:
+		tpl = "/var/lib/containers/storage/overlay-containers/$ID/userdata/ctr.log"
+	}
+	return strings.ReplaceAll(tpl, "$ID", id)
```
Comment (sgnn7) [inline pkg/logs/input/docker/launcher.go:310]: nit: Not a blocker but I think this helper would be easier to maintain in the future with an early exit and `fmt.Sprintfs`. Something like:

```go
if coreConfig.Datadog.GetBool("logs_config.use_podman_logs") {
    return fmt.Sprintf("/var/lib/containers/storage/overlay-containers/%s/userdata/ctr.log", id)
}

return filepath.Join(basePath, id, fmt.Sprintf("%s-json.log", id))
```

**Review by sgnn7 — APPROVED** [review]
@djmitche Just a couple of nits and a question but LGTM.

PS: It would be good to know why the integration test is detecting a data race though

File: pkg/config/config.go [hunk 3]
```diff
@@ -755,6 +755,11 @@ func InitConfig(config Config) {
 	config.BindEnvAndSetDefault("logs_config.auto_multi_line_default_match_timeout", 30) // Seconds
 	config.BindEnvAndSetDefault("logs_config.auto_multi_line_default_match_threshold", 0.48)
 
+	// If true, the agent looks for container logs in the location used by podman, rather
+	// than docker.  This is a temporary configuration parameter to support podman logs until
+	// a more substantial refactor of autodiscovery is made to determine this automatically.
+	config.BindEnvAndSetDefault("logs_config.use_podman_logs", false)
```
[the remark this answers is not in the document]
Comment (djmitche) [reply pkg/config/config.go:761]: Exactly.  We should have a more "automatic" way to detect this later.

File: pkg/logs/input/file/tailer.go (hunk 1 above)
Comment (djmitche) [reply pkg/logs/input/file/tailer.go:85]: I think I was just trying to keep with the nested-switch structure.  But, you're right, it's weird :)

**Comment by djmitche** [comment]
> PS: It would be good to know why the integration test is detecting a data race though

Yeah, I'm not sure what's up with that.  It's hard to tell with the main thread traceback failing.  Maybe it's to do with concurrent access to a context?  In any case, it seems limited to the pkg/metadata package.  I can't repro locally.

**Comment by gh123man** [comment]
Overall LGTM!
has there been any discussion around renaming `docker` to `container` or something similar since now our docker code handles more than one container runtime (and `DOCKER_HOST=unix:///run/podman/podman.sock` seems a bit strange) 

I know this would be a large breaking change - but I'm curious if there is a long term plan, or if this is good enough.

**Review by gh123man — APPROVED** [review]

**Comment by djmitche** [comment]
I think that may be part of the work regarding AD -- I'll find out.

**Pull request merged**